### PR TITLE
[Clang] Fix constexpr-ness on implicitly deleted destructors

### LINKED ISSUE #85550: clang-19 returns true for `__is_literal_type` on union with non trivial member starting with C++-23
Given the following code

```cpp
struct non_trivial {
    ~non_trivial();
};

union non_trivial_union {
    int i;
    non_trivial n;
};

static_assert(!__is_literal_type(non_trivial), "");
static_assert(!__is_literal_type(non_trivial_union), "");
```

clang-18.1.0 compiles fine while clang-19 fails the second `static_assert`.

[godbolt](https://godbolt.org/z/hPj9cGqhb)

Note that this does not happen in C++-20 mode [godbolt](https://godbolt.org/z/v94aM5fTj)

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
In C++20, a defaulted but implicitly deleted destructor is constexpr if and only if the class has no virtual base class. This hasn't been changed in C++23 by P2448R2.

Constexpr-ness on a deleted destructor affects almost nothing. The `__is_literal` intrinsic is related, while the corresponding
`std::is_literal_type(_v)` utility has been removed in C++20. A recently added example in `test/AST/ByteCode/cxx23.cpp` will become valid, and the example is already accepted by GCC.

Clang currently behaves correctly in C++23 mode, because the constexpr-ness on defaulted destructor is relaxed by P2448R2. But we should make similar relaxation for an implicitly deleted destructor.

Fixes #85550.

### CODE REVIEW

File: clang/include/clang/AST/CXXRecordDeclDefinitionBits.def [hunk 1]
```diff
@@ -81,6 +81,9 @@ FIELD(IsStandardLayout, 1, NO_MERGE)
 ///   member.
 FIELD(IsCXX11StandardLayout, 1, NO_MERGE)
 
+/// True when the class has a virtual base class.
+FIELD(HasVBases, 1, NO_MERGE)
```
Comment (AaronBallman) [inline clang/include/clang/AST/CXXRecordDeclDefinitionBits.def:85]: Do we need to add this here? Doesn't `DefinitionData::NumVBases` suffice?

File: clang/include/clang/AST/DeclCXX.h [hunk 2]
```diff
@@ -890,6 +890,13 @@ class CXXRecordDecl : public RecordDecl {
             needsOverloadResolutionForDestructor()) &&
            "destructor should not be deleted");
     data().DefaultedDestructorIsDeleted = true;
+    // C++23 [dcl.constexpr]p3.2:
+    //  if the function is a constructor or destructor, its class does not have
+    //  any virtual base classes.
+    // C++20 [dcl.constexpr]p5:
+    //   The definition of a constexpr destructor whose function-body is
+    //   [not = delete] shall additionally satisfy...
+    data().DefaultedDestructorIsConstexpr = !data().HasVBases;
```
Comment (AaronBallman) [inline clang/include/clang/AST/DeclCXX.h:899]: Can this be `data().NumVBases == 0`?

**Review by AaronBallman — COMMENTED** [review]
Thank you for the fix! Can you also add a release note to `clang/docs/ReleaseNotes.rst` so users know about the fix?

File: clang/include/clang/AST/CXXRecordDeclDefinitionBits.def (hunk 1 above)
Comment (frederick-vs-ja) [reply clang/include/clang/AST/CXXRecordDeclDefinitionBits.def:85]: Thanks for pointing this out!

**Review by cor3ntin — APPROVED** [review]
LGTM

**Pull request merged**